iterator.go

```diff
@@ -119,69 +114,133 @@ func (r RootLookuper) ExchangeWithServer(ctx context.Context, m *dns.Msg,
 }
 
 // Iterate is an iterative lookup implementation
-// revive:disable:cognitive-complexity
-// revive:disable:cyclomatic
 func (r RootLookuper) Iterate(ctx context.Context, name string,
 	qtype uint16, startAt string,
 ) (*dns.Msg, error) {
-	// revive:enable:cognitive-complexity
-	// revive:enable:cyclomatic
-	if startAt == "" {
-		startAt = pickRoot() + ":53"
+	req := new(dns.Msg)
+	req.SetQuestion(dns.Fqdn(name), qtype)
+	return r.IterateMsg(ctx, req, startAt)
+}
+
+// IterateMsg is an iterative exchange implementation
+func (r RootLookuper) IterateMsg(ctx context.Context, req *dns.Msg,
+	startAt string,
+) (*dns.Msg, error) {
+	var server string
+
+	switch {
+	case startAt != "":
+		server = startAt
+	case r.Start != "":
+		server = r.Start + ":53"
+	default:
+		server = pickRoot() + ":53"
+	}
+
+	msg, err := r.sanitizeRequest(req)
+	if err != nil {
+		return nil, err
 	}
-	server := startAt
-	name = dns.Fqdn(name)
-	msg := r.newMsgFromParts(name, qtype)
-	resp, err := r.ExchangeWithServer(ctx, msg, server)
+
+	return r.doIterate(ctx, msg, server)
+}
+
+func (RootLookuper) sanitizeRequest(req *dns.Msg) (*dns.Msg, error) {
+	if req == nil || len(req.Question) != 1 {
```

Comment:
This is not OK as per RFC. The RFC states that you can have more than 1 question in a dns message, you even have a 16 bit integer in the message header (QDCOUNT) that specifies how many questions you have in the message. Practice however discards all the questions except `req.Question[0]`. See https://github.com/miekg/dns/blob/836bea2f8dffa00b74af2d505c04e1cde3a9739b/types.go#L184

Reply:
Do you mean returning "not implemented" for Len(req.Question) != 1 instead of "bad request"?

Reply:
amended accordingly

Reply:
@karasz amended to handle the first question and ignore the rest